cbits/siphash.c

```diff
@@ -0,0 +1,262 @@
+	SIPROUND;
+	SIPROUND;
+    } else {
+	for (i = 0; i < d; i++)
+	    SIPROUND;
+    }
+    b = v0 ^ v1 ^ v2  ^ v3;
+    return b;
+}
+
+
+static inline uint64_t _siphash24(uint64_t k0, uint64_t k1, const u8 *str, size_t len)
+{
+    return _siphash(2, 4, k0, k1, str, len);
+}
+
+#if defined(__i386)
+# undef _siphash24
+
+static uint64_t (*_siphash24)(uint64_t k0, uint64_t k1, const u8 *, size_t);
+
+static void maybe_use_sse()
+    __attribute__((constructor));
+
+static void maybe_use_sse()
+{
+    uint32_t eax = 1, ebx, ecx, edx;
+
+    __asm volatile
+	("mov %%ebx, %%edi;" /* 32bit PIC: don't clobber ebx */
+	 "cpuid;"
+	 "mov %%ebx, %%esi;"
+	 "mov %%edi, %%ebx;"
+	 :"+a" (eax), "=S" (ebx), "=c" (ecx), "=d" (edx)
+	 : :"edi");
+
+#if defined(HAVE_SSE2)
+    if (edx & (1 << 26))
+	_siphash24 = hashable_siphash24_sse2;
+#if defined(HAVE_SSE41)
+    else if (ecx & (1 << 19))
+	_siphash24 = hashable_siphash24_sse41;
+#endif
+    else
+#endif
+	_siphash24 = plain_siphash24;
+}
+
+#endif
+
+/* ghci's linker fails to call static initializers. */
+static inline void ensure_sse_init()
+{
+#if defined(__i386)
+    if (_siphash24 == NULL)
+	maybe_use_sse();
+#endif
+}
+
+uint64_t hashable_siphash(int c, int d, uint64_t k0, uint64_t k1, const u8 *str, size_t len)
+{
+    return _siphash(c, d, k0, k1, str, len);
+}
+
+uint64_t hashable_siphash24(uint64_t k0, uint64_t k1, const u8 *str, size_t len)
+{
+    ensure_sse_init();
+    return _siphash24(k0, k1, str, len);
+}
+
+/* Used for ByteArray#s. We can't treat them like pointers in
+   native Haskell, but we can in unsafe FFI calls.
+ */
+uint64_t hashable_siphash24_offset(uint64_t k0, uint64_t k1,
+			      const u8 *str, size_t off, size_t len)
+{
+    ensure_sse_init();
+    return _siphash24(k0, k1, str + off, len);
+}
+
+static int _siphash_chunk(int c, int d, int buffered, uint64_t v[5],
+			  const u8 *str, size_t len, size_t totallen)
+{
+    uint64_t v0 = v[0], v1 = v[1], v2 = v[2], v3 = v[3], m, b;
+    const u8 *p, *end;
+    uint64_t carry = 0;
+    int i;
+
+    if (buffered > 0) {
+	int unbuffered = 8 - buffered;
+	int tobuffer = unbuffered > len ? len : unbuffered;
+	int shift = buffered << 3;
+
+	m = odd_read(str, tobuffer, v[4], shift);
+	str += tobuffer;
+	buffered += tobuffer;
+	len -= tobuffer;
+
+	if (buffered < 8)
+	    carry = m;
+	else {
+	    v3 ^= m;
+	    if (c == 2) {
+		SIPROUND;
+		SIPROUND;
+	    } else {
+		for (i = 0; i < c; i++)
+		    SIPROUND;
+	    }
+	    v0 ^= m;
+	    buffered = 0;
+	    m = 0;
+	}
+    }
+
+    for (p = str, end = str + (len & ~7); p < end; p += 8) {
+	m = peek_uint64_tle((uint64_t *) p);
+	v3 ^= m;
+	if (c == 2) {
+	    SIPROUND;
+	    SIPROUND;
+	} else {
+	    for (i = 0; i < c; i++)
+		SIPROUND;
+	}
+	v0 ^= m;
+    }
+
+    b = odd_read(p, len & 7, 0, 0);
+
+    if (totallen == -1) {
+	v[0] = v0;
+	v[1] = v1;
+	v[2] = v2;
+	v[3] = v3;
+	v[4] = b | carry;
+
+	return buffered + (len & 7);
+    }
+
+    b |= ((uint64_t) totallen) << 56;
```

Comment:
Isn't there a variant of siphash which doesn't force to include total length? This makes it uncomposable. Note how FNV-1 is hashing a stream, we just feed it data. (And also md5,sha1.sha256 etc.)

To repeat my previous comment, rebase on current `master` and try to contain the changes inside `LowLevel` module. It's not good that you need to reimplement how lazy test and bytestring hashable instances, you shouldn't. Int, bytearray/ptr buffer hash primitives should be enough.

Reply:
I eliminated this totallen, I think this is NIH code. (can't find it in the paper)

It's still not easily compose-able because siphash has an internal state of 4 64 bit words.
These need to be initialized with k0 and k1, but after this step it sortoff becomes like fnv if we keep track of that internal state separatly and you can fold over it (as long as you keep track of the state).
I still need to test this  out against that rechunck property test.

I think it's possible to express everything in terms of bytearray#, although I'm not sure how to get around the initilization/finilization steps.